### PR TITLE
Ensure server preference order in ALPN

### PR DESCRIPTION
**Pull request by leogomes**
Motivation:
With the current implementation the client protocol preference list
takes precedence over the one of the server, since the select method
will return the first item, in the client list, that matches any of the
protocols supported by the server. This violates the recommendation of
http://tools.ietf.org/html/rfc7301#section-3.2.

It will also fail with the current implementation of Chrome, which
sends back Extension application_layer_protocol_negotiation, protocols:
[http/1.1, spdy/3.1, h2-14]

Modifications:
Changed the protocol negotiator to prefer server’s list. Added a test
case that demonstrates the issue and that is fixed with the
modifications of this commit.

Result:
Server’s preference list is used.

### CODE REVIEW

**Comment by leogomes** [comment]
I've created the PR against 4.1, because that's the branch I'm currently working with. I can cherry-pick this on master, if needed. Just let me know.

**Comment by Scottmitch** [comment]
@leogomes - Oof...good catch :)  Can you explain the rational for changing `Set` -> `List`?  I'm guessing to keep object allocations down and lookup overhead is assumed to be low because the cardinality of the collection is typically small?  We may have to keep the interface for 4.0 (and maybe 4.1) @normanmaurer - WDYT?

**Comment by leogomes** [comment]
@Scottmitch - to respect the insert order I had two options, create a linked-hashset or keep the list that was already the interface exposed all the way up in the hierarchy. So, I reused the list using exactly the rationale you described, i.e. small cardinality, avoid object creation. I forgot to think about preserving the API :) and actually, it's so down-the-road that the list is converted into a set that it looked like a implementation detail, more than an API.

**Comment by normanmaurer** [comment]
@leogomes I think we need to keep the Set to not break the api. So could you change back to Set and ping me once done.

File: handler/src/test/java/io/netty/handler/ssl/JdkSslEngineTest.java [hunk 1]
```diff
@@ -279,6 +282,31 @@ public void testAlpnNoCompatibleProtocolsServerHandshakeFailure() throws Excepti
...
+            }
+            // Even the preferred application protocol appears second in the client's list, it will be picked
+            // because it's the first one on server's list.
+            JdkApplicationProtocolNegotiator clientApn = new JdkAlpnApplicationProtocolNegotiator(false, false,
+                FALLBACK_APPLICATION_LEVEL_PROTOCOL, PREFERRED_APPLICATION_LEVEL_PROTOCOL);
+            JdkApplicationProtocolNegotiator serverApn = new JdkAlpnApplicationProtocolNegotiator(true, true,
+                PREFERRED_APPLICATION_LEVEL_PROTOCOL, FALLBACK_APPLICATION_LEVEL_PROTOCOL);
+            mySetup(serverApn, clientApn);
+           assertNull(serverException);
+           runTest(PREFERRED_APPLICATION_LEVEL_PROTOCOL);
```
Comment (normanmaurer) [inline handler/src/test/java/io/netty/handler/ssl/JdkSslEngineTest.java]: Used 3 spaces and not 4 ?

Comment (leogomes) [inline handler/src/test/java/io/netty/handler/ssl/JdkSslEngineTest.java]: I have 2 spaces configured in Eclipse and probably added that by hand and missed 1 :) 
Funny thing is that the checkstyle plugin didn't complain at build-time. Anyway, I will start using IntelliJ for Netty development and grab the policy from your site ;-) I will fix that.

**Comment by trustin** [comment]
Looks good to me. Please feel free to cherry-pick once @leogomes fixes the indentation.

**Comment by leogomes** [comment]
@Scottmitch @normanmaurer PR updated with discussed changes.

File: handler/src/main/java/io/netty/handler/ssl/JdkApplicationProtocolNegotiator.java [hunk 2]
```diff
@@ -15,10 +15,11 @@
  */
 package io.netty.handler.ssl;
 
-import javax.net.ssl.SSLEngine;
 import java.util.List;
 import java.util.Set;
 
+import javax.net.ssl.SSLEngine;
```
Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/JdkApplicationProtocolNegotiator.java]: Was this change necessary?  If not please revert.

**Comment by Scottmitch** [comment]
@leogomes - Thanks for reverting back to set.  1 more comment then cherry-pick.

**Comment by normanmaurer** [comment]
@leogomes thanks.. just excluded the change that @Scottmitch commented on here and cherry-picked to 4.0 (034e0bd8d2bf3a2856a6627d317408c71216b52b),  4.1 (a97e413a659b8f576112eae803d3a7c6314e36f8) and master (4f13dee454079ce99c16c060d56875d25cbfb658)

**Comment by Scottmitch** [comment]
@leogomes Thanks for the patch!  Good work!

**Comment by leogomes** [comment]
@normanmaurer @Scottmitch - my pleasure! Netty is an awesome project and I'm happy when I'm able to contribute.
